packages/beacon-node/src/chain/opPools/singleAttestationPool.ts

```diff
@@ -0,0 +1,415 @@
+
+        const committeeSize = committeeInfo.committeeSize;
+        const attestationsByCommitteeValidatorIndex = committeeInfo.attestations;
+
+        const sameComitteeAttestations = new Map<CommitteeValidatorIndex, SingleAttestation>();
+        let newSeenEffectiveBalance = 0;
+        let newSeenAttesters = 0;
+        for (const notSeenCommitteeValidatorIndex of notSeenMembers) {
+          const attestation = attestationsByCommitteeValidatorIndex.get(notSeenCommitteeValidatorIndex);
+          if (attestation == null) {
+            // we don't have this missing SingleAttestation
+            continue;
+          }
+          newSeenEffectiveBalance += effectiveBalanceIncrements[attestation.attesterIndex];
+          newSeenAttesters++;
+          sameComitteeAttestations.set(notSeenCommitteeValidatorIndex, attestation);
+          // no need to search for the same notSeenCommitteeValidatorIndex in the next loop of attestation data root
+          notSeenMembers.delete(notSeenCommitteeValidatorIndex);
+        } // end looping through not seen committee members
+
+        if (sameComitteeAttestations.size === 0) {
+          // no missing SingleAttestations for this committeeIndex, expect this to happen a lot of times so not sure if we should track metrics here
+          continue;
+        }
+
+        const aggregatedAttestation = aggregateAttestations(sameComitteeAttestations, committeeIndex, committeeSize);
+        const attestationNonParticipation: AttestationNonParticipant = {
+          attestation: aggregatedAttestation,
+          newSeenEffectiveBalance,
+          newSeenAttesters,
+          notSeenCommitteeMembers: notSeenMembers,
+        };
+
+        sameAttDataCon.byCommittee.set(committeeIndex, attestationNonParticipation);
+        sameAttDataCon.totalNewSeenEffectiveBalance += attestationNonParticipation.newSeenEffectiveBalance;
+        sameAttDataCon.newSeenAttesters += attestationNonParticipation.newSeenAttesters;
+        sameAttDataCon.notSeenAttesters += attestationNonParticipation.notSeenCommitteeMembers.size;
+        sameAttDataCon.totalAttesters += committeeSize;
+      } // finish looping through all committee indices of the same attestation data
+
+      if (sameAttDataCon.byCommittee.size > 0 && sameAttDataCon.newSeenAttesters > 0) {
+        // we have at least one committee with new seen attesters
+        result.push(sameAttDataCon);
+      }
+    } // finish looping through all attestation data roots of the same slot
+
+    packedAttestationsMetrics?.returnedAttestations.set({inclusionDistance}, result.length);
+
+    // we bound returned AttestationsConsolidation in the consumer for each slot, so just return as many as possible
+    return result;
+  }
+
+  /**
+   * Remove any attestations with a slot lower than `current_slot - MAX_SLOTS_RETAINED`.
+   * Remove more slots until we have less than `MAX_ATTESTATIONS_RETAINED` attestations in the pool or at least `MIN_SLOTS_RETAINED` slots.
```

Comment:
Do we still have `MIN_SLOTS_RETAINED` mechanism? It looks like `prune()` will prune all attestations except those for current epoch and last epoch, and then prune down o `MAX_ATTESTATIONS_RETAINED`

Reply:
I did have `MIN_SLOTS_RETAINED` then removed since I switched to only store not-seen attestations
I don't see the use of it anymore. In unfinality time it's more important to keep the node stable than trying to find not-seen attestations to include in block

Reply:
>  In unfinality time it's more important to keep the node stable than trying to find not-seen attestations to include in block

For single attestations this strategy seems fine as it's not likely they will represent a lot of weight